### ISSUE TITLE
ENH: Add option to pass a single function to compute fun and jac in scipy.optimize.least_squares

### ISSUE THREAD

**Issue by frortega** [scipy/scipy#20826]
### Is your feature request related to a problem? Please describe.

it is not a problem by itself, but will allow improved efficiency in scipy.optimize.least_squares. Such function, asks for separate  callable objects to compute "fun" and "jac". The issue with such implementation is that several times computing "jac" requires expensive computations (inverse of some large dense matrix in my case) that is also performed as part of the computation of the output of "fun". Thus, i am performing expensive computations twice.

### Describe the solution you'd like.

The solution I would like is an option in scipy.optimize.least_squares to allow passing a single callable that returns both "fun" and "jac" value at each iteration of the nonlinear least squares problem.

**Comment by mdhaber** [comment]
There was a similar request for `minimize` that seemed to be resolved by https://github.com/scipy/scipy/issues/9265#issuecomment-423075191 and gh-17678. Would something similar resolve this request?

**Comment by frortega** [comment]
Thanks for pointing that out!.
I'll try to implement the ObjectiveWrapper  class for my specific problem and I'll get back here to comment on the matter as soon as possible.
Thanks again!

**Comment by andyfaff** [comment]
The `ObjectiveWrapper` class makes more sense (it's what the private `optimize._differentiable_functions.ScalarFunction` does) for caching. Bear in mind that the example given in #9265 isn't complete as it doesn't store the `x` value.

A similar thing can be done with function wrappers (I'll write an example today). @mdhaber I'm not sure #17678 will work, I was under the impression that `lru_cache` doesn't work for numpy arrays.

It makes sense to add this to documentation somewhere.

**Comment by mdhaber** [comment]
No, `lru_cache` won't work with arrays, but the idea of memorization is still applicable. Feel free to add a more general solution.

**Comment by andyfaff** [comment]
Something like:
```python
class Memoize:
    def __init__(self, fg):
        self.fg = fg
        self.cached_x = None
        self.cached_f = None
        self.cached_g = None

    def __call__(self, x):
        if np.array_equal(x, self.cached_x):
            pass
        else:
            self.cached_f, self.cached_g = self.fg(x)
            self.cached_x = x
        return self.cached_f, self.cached_g
        
    def fun(self, x):
        if np.array_equal(x, self.cached_x):
            pass
        else:
            self.cached_f, self.cached_g = self.fg(x)
            self.cached_x = x
        return self.cached_f

    def grad(self, x):
        if np.array_equal(x, self.cached_x):
            pass
        else:
            self.cached_f, self.cached_g = self.fg(x)
            self.cached_x = x
        return self.cached_g
```
should work.

**Comment by frortega** [comment]
Hi, 
thanks a lot for the solution, i got a (roughly measured) decrease in computation time of about 30%. 
[Here](https://github.com/frortega/EPIC_LS/blob/cfb92512cc099fa03aebf22e326bb8016ea59c04/EPIC_LS/EPIC/objective_fun_wrapper.py) is my implementation of the wrapper.
Thanks again!.

**Comment by dschmitz89** [comment]
@frortega : would you be interested to upstream your implementation to scipy itself? :)
